leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumObservationRegistryImpl.java

```diff
@@ -64,6 +66,7 @@
 
     private final List<ObservationRegistryListener> listeners = new CopyOnWriteArrayList<>();
     private final Map<KeyToken, Observation> observationsByToken = new ConcurrentHashMap<>();
+    private final Lock observationLock = new ReentrantLock();
```

Comment:
I think it is better to use a lock that is private to the class. Otherwise someone else could also synchronize on "this" and we might get a deadlock (more a theoretical and style discussion).

Of course we can also synchronize on a private object and use synchronized blocks.